src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -325,13 +329,56 @@ internal override void WriteWithQuotes(Utf8JsonWriter writer, T value, JsonSeria
 
             ulong key = ConvertToUInt64(value);
 
-            if (_nameCache.TryGetValue(key, out JsonEncodedText formatted))
+            JsonEncodedText formatted;
+            string original;
+
+            if (options.DictionaryKeyPolicy != null && !state.Current.IgnoreDictionaryKeyPolicy)
+            {
+                if (_dictionaryKeyPolicyCache.Value.TryGetValue(key, out formatted))
+                {
+                    writer.WritePropertyName(formatted);
+                    return;
+                }
+
+                original = value.ToString();
+
+                if (IsValidIdentifier(original))
+                {
+                    original = options.DictionaryKeyPolicy.ConvertName(original);
+
+                    if (original == null)
+                    {
+                        ThrowHelper.ThrowInvalidOperationException_NamingPolicyReturnNull(options.DictionaryKeyPolicy);
+                    }
+
+                    if (_dictionaryKeyPolicyCache.Value.Count < DictionaryKeyPolicyCacheSizeSoftLimit)
+                    {
+                        JavaScriptEncoder? encoder = options.Encoder;
+
+                        formatted = JsonEncodedText.Encode(original, encoder);
+
+                        writer.WritePropertyName(formatted);
+
+                        _dictionaryKeyPolicyCache.Value.TryAdd(key, formatted);
+                    }
+                    else
+                    {
+                        // We also do not create a JsonEncodedText instance here because passing the string
+                        // directly to the writer is cheaper than creating one and not caching it for reuse.
+                        writer.WritePropertyName(original);
+                    }
+
+                    return;
+                }
+            }
+
+            if (_nameCache.TryGetValue(key, out formatted))
             {
                 writer.WritePropertyName(formatted);
                 return;
             }
 
-            string original = value.ToString();
+            original = value.ToString();
```

Comment:
It feels like this method could be more concise. Could we refactor it such that we're performing this `IsValidIdentifier` if/else logic just once?